templates/template.py

```diff
@@ -0,0 +1,148 @@
+# {filename} for Unicode {unicode_version}
+# See https://github.com/ridiculousfish/widecharwidth/
+#
+# SHA1 file hashes:
+#  (
+#  the hashes for generate.py and the template are git object hashes,
+#  use `git log --all --find-object=<hash>` in the widecharwidth repository
+#  to see which commit they correspond to,
+#  or run `git hash-object` on the file to compare.
+#  The other hashes are simple `sha1sum` style hashes.
+#  )
+#
+#  generate.py:         {generate_hash}
+#  template.py:         {template_hash}
+#  UnicodeData.txt:     {unicode_hash}
+#  EastAsianWidth.txt:  {eaw_hash}
+#  emoji-data.txt:      {emoji_hash}
+
+__all__ = ["wcwidth", "Special"]
+
+from typing import Union
+from enum import Enum
+
+# Special width values
+class Special(Enum):
+    """The special values that wcwidth returns. Guaranteed to be negative.
+
+    nonprint = -1      : The character is not printable.
+    combining = -2     : The character is a zero-width combiner.
+    ambiguous = -3     : The character is East-Asian ambiguous width.
+    private_use = -4   : The character is for private use.
+    unassigned = -5    : The character is unassigned.
+    widened_in_9 = -6  : Width is 1 in Unicode 8, 2 in Unicode 9+.
+    non_character = -7 : The character is a noncharacter.
+    """
+
+    nonprint = -1  # The character is not printable.
+    combining = -2  # The character is a zero-width combiner.
+    ambiguous = -3  # The character is East-Asian ambiguous width.
+    private_use = -4  # The character is for private use.
+    unassigned = -5  # The character is unassigned.
+    widened_in_9 = -6  # Width is 1 in Unicode 8, 2 in Unicode 9+.
+    non_character = -7  # The character is a noncharacter.
+
+
+class Codepointrange:
+    """A list of ranges of codepoints."""
+
+    def __init__(self, *ranges):
+        self.ranges = sorted(ranges)
+
+    def __contains__(self, char):
+        left = 0
+        right = len(self.ranges) - 1
+        if char < self.ranges[0][0] or char > self.ranges[right][1]:
```

Comment:
nit, maybe more Pythonic as:

    if not self.ranges[0][0] <= char <= self.ranges[right][1]:
        return False